man/man8/zfs-load-key.8

```diff
@@ -289,6 +289,260 @@ would normally be. Since compression is applied before encryption datasets may
+the operation (lower-case) as its first argument,
+and the dataset name, if any, as its third argument.
+.El
+.Pp
+Example of a trivial provider:
+.Bd -literal -offset indent
+#!/bin/sh
+echo "$0" "$@"
+exec sh
+.Ed
+.Pp
+This will simply drop down into a shell, allowing the user to do whatever,
+for example running
+.Bd -literal -offset indent
+echo 'inheriting homomorphic to switc' >&3
+^D
+.Ed
+with
+.Sy keyformat
+set to
+.Sy raw
+will change the wrapping key to
+696e6865726974696e6720686f6d6f6d6f727068696320746f20737769746300.
+.Pp
+As a more contrived alternative, consider this provider:
+.Bd -literal -offset indent
+#!/bin/sh -ue
+
+if [ -z "$2" ]; then
+	echo "No dataset name (zfs-create?)" >&2
+	exit 2
+fi
+
+first="$(zfs get -Hpo value 'xyz.nabijaczleweli:bad-provider.salt' "$2")"
+second="$(zfs get -Hpo value 'xyz.nabijaczleweli:bad-provider.saltmp' "$2")"
+
+
+hash() {
+	pass=""
+	while [ -z "$pass" ]; do
+		read -rp "Passphrase: " pass
+	done
+
+	if command -v sha256 >/dev/null; then
+		sha256 -qs "$pass$1"
+	else
+		echo -n "$pass$1" | sha256sum | awk '{print $1}'
+	fi
+}
+
+
+case "$1" in
+	load)
+		if [ "$first" = "-" ] && [ "$second" != "-" ]; then
+			zfs set 'xyz.nabijaczleweli:bad-provider.salt'="$first" "$2"
+			zfs inherit 'xyz.nabijaczleweli:bad-provider.saltmp' "$2"
+			first="$second"
+			second="-"
+		elif [ "$first" = "-" ] && [ "$second" = "-" ]; then
+			echo "No state?" >&2
+			exit 3
+		elif [ "$first" != "-" ] && [ "$second" != "-" ]; then
+			which=""
+			while [ "$which" != "f" ] && [ "$which" != "s" ]; do
+				read -rp "Both states present! Select which one to use [fs]: " which
+			done
+			if [ "$which" = "f" ]; then
+				echo "If this is the right key, run '$0 shift $2' afterward"
+			else
+				echo "If this is the right key, run '$0 cancel $2' afterward"
+				first="$second"
+			fi
+			second="-"
+		fi
+
+		hash "$first" >&3
+		;;
+	new)
+		if [ "$second" != "-" ]; then
+			echo "Second slot occupied? Run 'zfs load-key [-n] $2' to resolve this." >&2
+			exit 4
+		fi
+
+		second="$(tr -cd '[:alnum:]' < /dev/urandom | dd bs=128 count=1 status=none)"
+		zfs set 'xyz.nabijaczleweli:bad-provider.saltmp'="$second" "$2"
+
+		hash "$second" >&3
+		;;
+	shift)
+		if [ "$first" != "-" ]; then
+			zfs inherit 'xyz.nabijaczleweli:bad-provider.salt' "$2"
+			first="-"
+		fi
+
+		zfs set 'xyz.nabijaczleweli:bad-provider.salt'="$second" "$2"
+		zfs inherit 'xyz.nabijaczleweli:bad-provider.saltmp' "$2"
+		first="$second"
+		second="-"
+		;;
+	unshift)
+		if [ "$second" != "-" ]; then
+			zfs inherit 'xyz.nabijaczleweli:bad-provider.saltmp' "$2"
+			second="-"
+		fi
+
+		zfs set 'xyz.nabijaczleweli:bad-provider.saltmp'="$first" "$2"
+		zfs inherit 'xyz.nabijaczleweli:bad-provider.salt' "$2"
+		second="$first"
+		first="-"
+		;;
+	cancel)
+		if [ "$second" != "-" ]; then
+			zfs inherit 'xyz.nabijaczleweli:bad-provider.saltmp' "$2"
+			second="-"
+		fi
+		;;
+	*)
+		echo "Unknown op $1" >&2
+		exit 5
+		;;
+esac
```

Comment:
I'm not sure a 100 line code example fits in the man page. (Note however that I'm also not a zfs maintainer or dev)

Reply:
Oh, I'm quite sure it doesn't, but I wasn't sure where to put it so it made sense. I hope to be actionably told off by a maintainer to use some specific other useful place.